File: Swift/862.和至少为-k-的最短子数组.swift
/*
 * @lc app=leetcode.cn id=862 lang=swift
 *
 * [862] 和至少为 K 的最短子数组
 */

// @lc code=start
class Solution {
    func shortestSubarray(_ nums: [Int], _ k: Int) -> Int {
        let n = nums.count
        var preSums = [Int](repeating: 0, count: n + 1)
        for i in 0..<n {
            preSums[i + 1] = preSums[i] + nums[i]
        }
        var ans = n + 1
        var q = [Int]()
        for i in 0...n {
            let curSum = preSums[i]
            while !q.isEmpty && curSum - preSums[q.first!] >= k {
                ans = min(ans, i - q.removeFirst())
            }
            while !q.isEmpty && preSums[q.last!] >= curSum {
                q.removeLast()
            }
            q.append(i)
        }
        return ans < n + 1 ? ans : -1
    }
}
// @lc code=end

func main() {
    var nums: [Int]
    var k: Int
    var res: Int
    
    /**
     输入：nums = [1], k = 1
     输出：1
     */
    nums = [1]
    k = 1
    res = 1
    assert(res == Solution().shortestSubarray(nums, k))
    
    /**
     输入：nums = [1,2], k = 4
     输出：-1
     */
    nums = [1,2]
    k = 4
    res = -1
    assert(res == Solution().shortestSubarray(nums, k))
    
    /**
     输入：nums = [2,-1,2], k = 3
     输出：3
     */
    nums = [2,-1,2]
    k = 3
    res = 3
    assert(res == Solution().shortestSubarray(nums, k))
}

File: Cpp/862.和至少为-k-的最短子数组.cpp
/*
 * @lc app=leetcode.cn id=862 lang=cpp
 *
 * [862] 和至少为 K 的最短子数组
 */

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// @lc code=start
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        int n = (int)nums.size();
        vector<long> preSums(n + 1, 0);
        for (int i = 0; i < n; ++i) {
            preSums[i + 1] = preSums[i] + nums[i];
        }
        int ans = n + 1;
        deque<int> q;
        for (int i = 0; i <= n; ++i) {
            long curSum = preSums[i];
            while (!q.empty() && curSum - preSums[q.front()] >= k) {
                ans = min(ans, i - q.front());
                q.pop_front();
            }
            while (!q.empty() && preSums[q.back()] >= curSum) {
                q.pop_back();
            }
            q.push_back(i);
        }
        return ans < n + 1 ? ans : -1;
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
    
    vector<int> nums;
    int k;
    int res;
    
    /**
     输入：nums = [1], k = 1
     输出：1
     */
    nums = {1};
    k = 1;
    res = 1;
    assert(res == Solution().shortestSubarray(nums, k));
    
    /**
     输入：nums = [1,2], k = 4
     输出：-1
     */
    nums = {1,2};
    k = 4;
    res = -1;
    assert(res == Solution().shortestSubarray(nums, k));
    
    /**
     输入：nums = [2,-1,2], k = 3
     输出：3
     */
    nums = {2,-1,2};
    k = 3;
    res = 3;
    assert(res == Solution().shortestSubarray(nums, k));
    
    cout << "OK~" << endl;
    
    return 0;
}
